### PR TITLE
Playing around with simplified prompt interface

### PR DESCRIPTION
**Pull request by eob**
## Description

This PR introduces our second PluginInstance subclass -- `PromptGenerationPluginInstance`.

It changes no semantics of how plugins, tagging, or existing LLM plugins work; it only acts as a convenience wrapper around PluginInstances that happen to be associated with an LLM tagger.

* It adds a `generate(prompt, prompt_variables)` method which synchronously returns a string.
* It adds an output sanitization method.

## Example Usage

Using this new subclass, one can invoke an LLM with two lines of code and two lines of data:

```
PROMPT = "Greet {name} as if he were a {relation}."
VARIABLES = { "name": "Ted", "relation": "friend" }

llm = Steamship.use('prompt-generation-default', config={ "temperature": 0.9 })
greeting = llm.generate(PROMPT, VARIABLES)
```

## Implementation

* The implementation of `generate` is copy-pasted directly from our current prompt-api template. The intention is to simplify our code with as small as possible a change in present-day mechanics.
* The knowledge of when to use this subclass is hard-coded into our client, as we do with embedding indices, which seems OK since at most there are only 4-5 LLM providers that matter at present.

### CODE REVIEW

File: src/steamship/client/steamship.py [hunk 1]
```diff
@@ -144,7 +153,10 @@ def _instance_use(
             fetch_if_exists=fetch_if_exists,
         )
 
-        return instance
+    @staticmethod
+    def register_plugin_instance_subclass(plugin_handle: str, plugin_class: PluginInstance):
```
Comment (EniasCailliau) [inline src/steamship/client/steamship.py:157]: If we're only planning to use this method during testing I would directly access the static variable in our test scripts instead of using a static function. 

In the tests this would look like: 

```
Steamship._PLUGIN_INSTANCE_SUBCLASS_OVERRIDES[plugin_handle] = plugin_class
```

File: src/steamship/data/plugin/prompt_generation_plugin_instance.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+
+class PromptGenerationPluginInstance(PluginInstance):
+    """An instance of a configured prompt completion service such as GPT-3.
+
+    The `generate` method synchronously invokes the prompt against a set of variables that parameterize it.
+    The return value is a single string.
+
+    Example Usage:
+       llm = Steamship.use('prompt-generation-default', config={ "temperature": 0.9 })
```
Comment (EniasCailliau) [inline src/steamship/data/plugin/prompt_generation_plugin_instance.py:16]: Does it make sense to use `gpt3` here instead of `prompt-generation-default`?

**Review by EniasCailliau — APPROVED** [review]
Looks good!

File: src/steamship/client/steamship.py (hunk 1 above)
Comment (dkolas) [reply src/steamship/client/steamship.py:157]: +1 to this- I think more thought would be necessary to figure out an extension pattern for user-submitted plugins.

File: src/steamship/data/plugin/prompt_generation_plugin_instance.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        #
+        # To learn about the Steamship data model, please consult our the docs:
+        # https://docs.steamship.com/workspaces/data_model/index.html
+        file = File.create(self.client, blocks=[Block.CreateRequest(text=prompt_text)])
+
+        # This requests generation from the parameterized prompt. Tagging with our prompt generator
+        # plugin will result in a new tag that contains the generated output.
+        # We `wait()` because generation of text is done asynchronously and may take a few moments
+        # (somewhat depending on the complexity of your prompt).
+        tag_task = file.tag(plugin_instance=self.handle)
```
Comment (dkolas) [inline src/steamship/data/plugin/prompt_generation_plugin_instance.py:44]: If we're abstracting away seeing the file created, would it make more sense for this to use the inline tagging method?

tag_task = `self.tag(doc=prompt_text)`

File: src/steamship/data/plugin/prompt_generation_plugin_instance.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        # Here, we iterate through the content blocks associated with a file
+        # as well as any tags on that content to find the generated text.
+        #
+        # The Steamship data model provides flexible content organization,
+        # storage, and lookup. Read more about the data model via:
+        # https://docs.steamship.com/workspaces/data_model/index.html
+        for text_block in tag_task.output.file.blocks:
+            for block_tag in text_block.tags:
+                if block_tag.kind == TagKind.GENERATION:
+                    return self._clean_output(block_tag.value[TagValueKey.STRING_VALUE])
```
Comment (dkolas) [inline src/steamship/data/plugin/prompt_generation_plugin_instance.py:58]: If this isn't going to be in user code, I think there should be a flag on generate (default true) on whether to do this.  It's not clear to me it's the right behavior for all applications.

File: tests/steamship_tests/plugin/integration/test_e2e_prompt_plugin.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+        # Now we just invoke use_plugin.
+        # Note: this just applies Steamship.use_plugin with the same testing environment as the unit tests.
+        with steamship_use_plugin(plugin.handle, delete_workspace=True) as llm:
+            # And we can run a prompt. Covering some bases to encode edge cases and their expected behavior.
+
+            # Prompt vars: 1, Provided vars: 1
+            assert llm.generate("Hello, {name}!", {"name": "world"}) == "Hello, world!"
+            # Prompt vars: 0, Provided vars: 1
+            assert llm.generate("Hello, World!", {"name": "world"}) == "Hello, World!"
```
Comment (dkolas) [inline tests/steamship_tests/plugin/integration/test_e2e_prompt_plugin.py:33]: Can this throw an error?

**Review by dkolas — APPROVED** [review]
Looks good - a few comments / suggestions

File: src/steamship/client/steamship.py (hunk 1 above)
Comment (eob) [reply src/steamship/client/steamship.py:157]: Will do!

File: src/steamship/data/plugin/prompt_generation_plugin_instance.py (hunk 3 above)
Comment (eob) [reply src/steamship/data/plugin/prompt_generation_plugin_instance.py:44]: Done!

File: src/steamship/data/plugin/prompt_generation_plugin_instance.py (hunk 4 above)
Comment (eob) [reply src/steamship/data/plugin/prompt_generation_plugin_instance.py:58]: Ooh. Good point.

File: tests/steamship_tests/plugin/integration/test_e2e_prompt_plugin.py (hunk 5 above)
Comment (eob) [reply tests/steamship_tests/plugin/integration/test_e2e_prompt_plugin.py:33]: I think so -- if the tag shape is unexpected. Should we softly handle it w/ an empty string result?
